### PR TITLE
feat: introduce flash attention with variable length

### PR DESCRIPTION
**Pull request by jorgeantonio21**

### CODE REVIEW

File: csrc/build.rs [hunk 1]
```diff
@@ -74,7 +74,10 @@ const KERNEL_FILES: [&str; 65] = [
 
 fn main() -> Result<()> {
     println!("cargo:rerun-if-changed=build.rs");
-    for kernel_file in KERNEL_FILES.iter() {
+    for kernel_file in KERNEL_FILES
+        .iter()
+        .chain(std::iter::once(&"kernels/flash_api.cu"))
```
Comment (Cifko) [inline csrc/build.rs:79]: the `flash_api.cu` is already in the `KERNEL_FILES`

File: csrc/build.rs [hunk 2]
```diff
@@ -91,19 +94,10 @@ fn main() -> Result<()> {
 
     let out_dir = PathBuf::from(std::env::var("OUT_DIR").context("OUT_DIR not set")?);
     let build_dir = match std::env::var("ATOMA_FLASH_ATTN_BUILD_DIR") {
```
Comment (Cifko) [inline csrc/build.rs:96]: will we ever use the custom build dir?

File: csrc/kernels/flash_api.cu [hunk 3]
```diff
@@ -71,7 +72,13 @@ extern "C" void run_mha(
     int window_size_left,
     int window_size_right,
     float softcap,
+<<<<<<< HEAD
```
Comment (Cifko) [inline csrc/kernels/flash_api.cu:75]: this file has many unresolved merge issues

File: csrc/kernels/flash_fwd_launch_template.h [hunk 4]
```diff
@@ -207,11 +207,17 @@ void run_mha_fwd_hdim64(Flash_fwd_params &params, cudaStream_t stream) {
 template<typename T, bool Is_causal>
 void run_mha_fwd_hdim96(Flash_fwd_params &params, cudaStream_t stream) {
     constexpr static int Headdim = 96;
+<<<<<<< HEAD
```
Comment (Cifko) [inline csrc/kernels/flash_fwd_launch_template.h:210]: another one, I will continue review once this is fixed.

File: csrc/build.rs (hunk 1 above)
Comment (jorgeantonio21) [reply csrc/build.rs:79]: Good catch !

File: csrc/build.rs (hunk 2 above)
Comment (jorgeantonio21) [reply csrc/build.rs:96]: I think that will be the case, if you want to save the cuda compiled binaries without having to compromise your `rust` build.

File: csrc/kernels/flash_api.cu (hunk 3 above)
Comment (jorgeantonio21) [reply csrc/kernels/flash_api.cu:75]: it didn't save when I changed these

File: csrc/src/lib.rs [hunk 5]
```diff
@@ -280,6 +289,7 @@ impl FlashAttention {
                 /* window_size_right */ window_size_right,
                 /* softcap */ softcap,
                 false,
```
Comment (Cifko) [inline csrc/src/lib.rs:291]: all of the parameters above have the argument name in the comment, maybe add these two as well?

File: csrc/src/lib.rs [hunk 6]
```diff
@@ -357,6 +366,1038 @@ pub fn flash_attn(
...
+                /* seqlen_k */ self.max_seqlen_k as u32,
+                /* seqlen_q_rounded */ seqlen_q_rounded as u32,
+                /* seqlen_k_rounded */ seqlen_k_rounded as u32,
+                /* is_bf16 */ is_bf16,
+                /* is_causal */ is_causal,
+                /* window_size_left */ window_size_left,
+                /* window_size_right */ window_size_right,
+                /* softcap */ softcap,
+                /* unpadded_lse */ true,
+                false,
```
Comment (Cifko) [inline csrc/src/lib.rs:1000]: add the comment with argument name here as well

**Review by Cifko — APPROVED** [review]

File: csrc/src/lib.rs (hunk 5 above)
Comment (jorgeantonio21) [reply csrc/src/lib.rs:291]: addressed

**Pull request merged**